File: simple_xml_builder.rb
# Hit CTRL+ALT+N to run inside VS Code.

# Try it on the browser:
# https://try.ruby-lang.org/playground/#code=class+Cursor%0A++attr_reader+%3Aposition%0A++def+initialize%3B+%40position+%3D+%5B0%2C+0%5D%3B+end%0A++def+move(%26block)%3B+instance_exec+%26block%3B+end%0A++def+up(steps+%3D+1)%3B+position%5B0%5D+%2B%3D+steps%3B+end%0A++def+down(steps+%3D+1)%3B+position%5B0%5D+-%3D+steps%3B+end%0A++def+right(steps+%3D+1)%3B+position%5B1%5D+%2B%3D+steps%3B+end%0A++def+left(steps+%3D+1)%3B+position%5B1%5D+-%3D+steps%3B+end%0Aend%0A%0Acursor+%3D+Cursor.new%0A%0Acursor.move+do%0A++up%0A++down+2%0A++left+3%0A++right+4%0Aend%0A%0A%23+Should+display+%5B-1%2C+1%5D.%0Ap+cursor.position%0A&engine=opal

# class Cursor
#   attr_reader :position
#   def initialize; @position = [0, 0]; end
#   def move(&block); instance_exec &block; end
#   def up(steps = 1); position[0] += steps; end
#   def down(steps = 1); position[0] -= steps; end
#   def right(steps = 1); position[1] += steps; end
#   def left(steps = 1); position[1] -= steps; end
# end

# cursor = Cursor.new

# cursor.move do
#   up
#   down 2
#   left 3
#   right 4
# end

# # Should display [-1, 1]
# p cursor.position

class Array
  private

  def method_missing(name, *args, &block)
    if all? { |item| item.respond_to? name }
      map { |item| item.send name, *args, &block }
    elsif any? { |item| item.respond_to? name }
      raise "Sorry, but all items need to respond to method `#{name}` to allow it to be called over the whole collection"
    else
      super
    end
  end
end

class SimpleXMLBuilder
  attr_reader :root_element
  private attr_reader :current_depth, :elements

  # Undefine most inherited instance methods (e.g. dup, to_s, then, display, taint, class, clone
  # etc), in order to allow them to be used as XML tag names.
  (instance_methods - instance_methods(false))
    .filter do |method|
      # Exclude a few selected methods, ones that start and end with `__` or which contain at
      # least one non-word character (e.g. ?, ! etc).
      !(
        %i[object_id instance_exec].include?(method) || # Remember to add :rand when running in Opal.
          method.to_s =~ /^__\w+__$/ || method.to_s =~ /\W/
      )
    end
    .each { |method| undef_method method }

  def initialize(&block)
    build(&block) if block
  end

  def build(&block)
    reset
    instance_exec &block
  end

  def str
    root_element.str
  end

  private

  def reset
    @elements = []
    @current_depth = 0
  end

  def method_missing(name, *args, &block)
    additional_params =
      args[0].is_a?(Hash) ? [args[0], nil] : [args[1], args[0]]

    new_element = XMLElement.new(name, *additional_params)

    @root_element ||= new_element
    elements << [current_depth, new_element]

    return unless block

    process_children_elements do
      previous_elements = elements.clone

      result = block.call

      # Any new elements added?
      if elements.size > previous_elements.size
        added_elements =
          (elements - previous_elements).filter_map do |(depth, element)|
            element if depth == current_depth
          end

        new_element.children = added_elements
      else
        # Assume the block returned an scalar value (number, string etc) instead.
        new_element.children = result
      end
    end
  end

  def process_children_elements(&block)
    @current_depth += 1
    block.call
    @current_depth -= 1
  end
end

class XMLElement
  INDENTATION_SPACING = 2

  attr_reader :name, :attrs
  attr_accessor :children

  def initialize(name, attrs = {}, children = [])
    @name = name
    @attrs = attrs
    @children = children
  end

  def str(depth = 0)
    if children
      children_as_string =
        if children.is_a?(Array)
          children.str(depth + 1).join("\n")
        else
          [indentation(depth + 1), children].join
        end

      [
        indentation(depth) + opening_tag(attrs),
        children_as_string,
        indentation(depth) + closing_tag,
      ].join("\n")
    else
      indentation(depth) + opening_tag(attrs, self_closing: true)
    end
  end

  private

  def indentation(depth)
    ' ' * INDENTATION_SPACING * depth
  end

  def opening_tag(attrs, self_closing: false)
    attrs_as_string = attrs ? ' ' + attrs_as_key_value_pairs_string : ''

    ['<', name, attrs_as_string, self_closing ? ' /' : '', '>'].join
  end

  def closing_tag
    ['</', name, '>'].join
  end

  def attrs_as_key_value_pairs_string
    attrs.map { |(k, v)| %Q(#{k}="#{v}") }.join(' ')
  end
end

xml =
  SimpleXMLBuilder.new do
    document type: 'xml', use: 'example' do
      description 'This is an example of using SimpleXMLBuilder2.'
      next_meeting date: Time.now + 100_000 do
        agenda 'Nothing of importance will be decided.'
        clearance level: 'classified'
      end
    end
  end

puts xml.str

# Produces:
#
# <document type="xml" use="example">
#   <description>
#     This is an example of using SimpleXMLBuilder.
#   </description>
#   <next_meeting date="2022-12-28 07:47:15 -0300">
#     <agenda>
#       Nothing of importance will be decided.
#     </agenda>
#     <clearance level="classified" />
#   </next_meeting>
# </document>

xml =
  SimpleXMLBuilder.new do
    html lang: 'en-US' do
      head do
        title 'Just a moment...'
        link href: '/cdn-cgi/styles/challenges.css', rel: 'stylesheet'
      end
      body class: 'no-js' do
        div class: 'main-wrapper', role: 'main'
        div class: 'main-content' do
          h1 class: 'zone-name-title h1' do
            img class: 'heading-favicon', src: '/favicon.ico'
          end
          h2 class: 'h2', id: 'challenge-running' do
            'Checking if the site connection is secure'
          end
          noscript do
            div id: 'challenge-error-title' do
              div class: 'h2' do
                span class: 'icon-wrapper' do
                  div class: 'heading-icon warning-icon'
                  span 'Enable JavaScript and cookies to continue',
                       id: 'challenge-error-text'
                end
              end
            end
          end
        end
      end
    end
  end

puts
puts xml.str

# Produces:
#
# <html lang="en-US">
#   <head>
#     <title>
#       Just a moment...
#     </title>
#     <link href="/cdn-cgi/styles/challenges.css" rel="stylesheet" />
#   </head>
#   <body class="no-js">
#     <div class="main-wrapper" role="main" />
#     <div class="main-content">
#       <h1 class="zone-name-title h1">
#         <img class="heading-favicon" src="/favicon.ico" />
#       </h1>
#       <h2 class="h2" id="challenge-running">
#         Checking if the site connection is secure
#       </h2>
#       <noscript>
#         <div id="challenge-error-title">
#           <div class="h2">
#             <span class="icon-wrapper">
#               <div class="heading-icon warning-icon" />
#               <span id="challenge-error-text">
#                 Enable JavaScript and cookies to continue
#               </span>
#             </span>
#           </div>
#         </div>
#       </noscript>
#     </div>
#   </body>
# </html>

xml = SimpleXMLBuilder.new

xml.build do
  report do
    name 'Annual Report'
    xml.class 'Class of 94' # Notice the explicit use of the `xml` receiver, since `class` is a reserved word.
    12.times do |i|
      amounts month: i + 1 do
        expenses rand(1000)
        revenue { rand(1000) }
      end
    end
  end
end

puts
puts xml.str

# Produces:
#
# <report>
#   <name>
#     Annual report
#   </name>
#   <month index="1">
#     <expenses>
#       812
#     </expenses>
#     <revenue>
#       259
#     </revenue>
#   </month>
#   <month index="2">
#     <expenses>
#       898
#     </expenses>
#     <revenue>
#       534
#     </revenue>
#   </month>
#   <month index="3">
#     <expenses>
#       479
#     </expenses>
#     <revenue>
#       844
#     </revenue>
#   </month>
#   <month index="4">
#     <expenses>
#       870
#     </expenses>
#     <revenue>
#       575
#     </revenue>
#   </month>
#   <month index="5">
#     <expenses>
#       239
#     </expenses>
#     <revenue>
#       472
#     </revenue>
#   </month>
#   <month index="6">
#     <expenses>
#       457
#     </expenses>
#     <revenue>
#       669
#     </revenue>
#   </month>
#   <month index="7">
#     <expenses>
#       387
#     </expenses>
#     <revenue>
#       425
#     </revenue>
#   </month>
#   <month index="8">
#     <expenses>
#       986
#     </expenses>
#     <revenue>
#       866
#     </revenue>
#   </month>
#   <month index="9">
#     <expenses>
#       975
#     </expenses>
#     <revenue>
#       701
#     </revenue>
#   </month>
#   <month index="10">
#     <expenses>
#       245
#     </expenses>
#     <revenue>
#       87
#     </revenue>
#   </month>
#   <month index="11">
#     <expenses>
#       785
#     </expenses>
#     <revenue>
#       135
#     </revenue>
#   </month>
#   <month index="12">
#     <expenses>
#       218
#     </expenses>
#     <revenue>
#       6
#     </revenue>
#   </month>
# </report>

File: simple_xml_builder.ts
type XMLBuilderPrimitiveType = string | number | boolean

type XMLBuilderFnType = (
  xml: SimpleXMLBuilder
) => void | XMLBuilderPrimitiveType

class SimpleXMLBuilder {
  public rootElement?: XMLElement
  private elements!: [number, XMLElement][]
  private currentDepth!: number

  constructor(fn: (xml: SimpleXMLBuilder) => void) {
    if (fn) {
      this.build(fn)
    } else {
      this.reset()
    }
  }

  build(fn: (xml: SimpleXMLBuilder) => void) {
    this.reset()
    fn(this.createProxy())
  }

  str(): string | undefined {
    return this.rootElement?.str()
  }

  /////////////////////
  // Private methods //
  /////////////////////

  reset(): void {
    this.rootElement = undefined
    this.elements = []
    this.currentDepth = 0
  }

  createProxy(): this {
    const proxy: this = new Proxy(this, {
      get(target, name: string) {
        return (
          arg1?: object | XMLBuilderPrimitiveType | XMLBuilderFnType,
          arg2?: object | XMLBuilderFnType
        ) => {
          let additionalXMLElementParams: [
            undefined | object,
            undefined | XMLBuilderPrimitiveType
          ]
          let fn: undefined | XMLBuilderFnType

          switch (typeof arg1) {
            case 'object':
              additionalXMLElementParams = [arg1, undefined]

              if (!(arg2 instanceof Function || arg2 === undefined))
                throw new Error('Invalid type for arg2')

              fn = arg2
              break
            case 'function':
              additionalXMLElementParams = [undefined, undefined]

              fn = arg1 as XMLBuilderFnType
              break
            default: // arg1 is probably a primitive value.
              if (arg2 instanceof Function)
                throw new Error('Invalid type for arg2')

              additionalXMLElementParams = [arg2, arg1]

              fn = undefined
          }

          const newElement = new XMLElement(name, ...additionalXMLElementParams)

          target.elements.push([target.currentDepth, newElement])

          target.rootElement = target.rootElement || newElement

          if (!fn) return

          target.processChildrenElements(() => {
            const previousElements = [...target.elements]
            const result = fn?.(proxy)

            // Any new elements added?
            if (target.elements.length > previousElements.length) {
              const isSameItem = (
                [depthA, elementA]: [number, XMLElement],
                [depthB, elementB]: [number, XMLElement]
              ) => depthA === depthB && elementA === elementB

              const addedElements = arraysDiff(
                target.elements,
                previousElements,
                isSameItem
              )
                .filter(([depth, _]) => depth === target.currentDepth)
                .map(([_, element]) => element)

              newElement.children = addedElements
            } else {
              if (result === undefined) throw new Error('result is undefined')

              // Assume the block returned a primitive value (number, string etc) instead.
              newElement.children = result
            }
          })
        }
      },
    })

    return proxy
  }

  processChildrenElements(fn: () => void): void {
    this.currentDepth++
    fn()
    this.currentDepth--
  }
}

const INDENTATION_SPACING = 2

class XMLElement {
  constructor(
    public name: string,
    public attrs?: object,
    public children?: XMLBuilderPrimitiveType | XMLElement[]
  ) {}

  str(depth: number = 0): string {
    if (this.children !== undefined) {
      const childrenAsString = Array.isArray(this.children)
        ? this.children.map(child => child.str(depth + 1)).join('\n')
        : this.indentation(depth + 1) + this.children.toString()

      return [
        this.indentation(depth) + this.openingTag(),
        childrenAsString,
        this.indentation(depth) + this.closingTag(),
      ].join('\n')
    } else {
      return this.indentation(depth) + this.openingTag(true)
    }
  }

  indentation(depth: number): string {
    return ' '.repeat(INDENTATION_SPACING * depth)
  }

  openingTag(selfClosing: boolean = false): string {
    const attrsAsString =
      this.attrs !== undefined ? ' ' + this.attrsAsKeyValuePairsString() : ''

    return `<${this.name}${attrsAsString}${selfClosing ? ' /' : ''}>`
  }

  closingTag(): string {
    return `</${this.name}>`
  }

  attrsAsKeyValuePairsString(): string {
    return Object.entries(this.attrs!)
      .map(([k, v]) => `${k}="${v}"`)
      .join(' ')
  }
}

///////////////////////
// Utility functions //
///////////////////////

const arraysDiff = <T>(
  left: T[],
  right: T[],
  comparisonFn: (left: T, right: T) => boolean
) =>
  left.filter(
    leftItem => !right.find(rightItem => comparisonFn(leftItem, rightItem))
  )

function times(count: number, fn: Function): number | number[] {
  const result: number[] = []

  for (let i = 0; i < count; i++) {
    if (fn === undefined) {
      result.push(i)
    } else {
      fn(i)
    }
  }

  return fn === undefined ? result : count
}

function rand(limit: number): number {
  return Math.trunc(Math.random() * limit)
}

const builder: SimpleXMLBuilder = new SimpleXMLBuilder(xml => {
  // @ts-ignore
  xml.document({ type: 'xml', use: 'example' }, () => {
    // @ts-ignore
    xml.description('This is an example of using SimpleXMLBuilder2.')
    // @ts-ignore
    xml.next_meeting({ date: new Date() }, () => {
      // @ts-ignore
      xml.agenda('Nothing of importance will be decided.')
      // @ts-ignore
      xml.clearance({ level: 'classified' })
    })
  })
})

console.log(builder.str())

/*
<document type="xml" use="example">
  <description>
    This is an example of using SimpleXMLBuilder2.
  </description>
  <next_meeting date="Thu Jan 05 2023 01:15:29 GMT-0300 (Brasilia Standard Time)">
    <agenda>
      Nothing of importance will be decided.
    </agenda>
    <clearance level="classified" />
  </next_meeting>
</document>
*/

builder.build(xml => {
  // @ts-ignore
  xml.html({ lang: 'en-US' }, () => {
    // @ts-ignore
    xml.head(() => {
      // @ts-ignore
      xml.title('Just a moment...')
      // @ts-ignore
      xml.link('Some link', {
        href: '/cdn-cgi/styles/challenges.css',
        rel: 'stylesheet',
      })
    })
    // @ts-ignore
    xml.body({ class: 'no-js' }, () => {
      // @ts-ignore
      xml.div({ class: 'main-wrapper', role: 'main' }, () => 'Some div')
      // @ts-ignore
      xml.br()
    })
  })
})

/*
<html lang="en-US">
  <head>
    <title>
      Just a moment...
    </title>
    <link href="/cdn-cgi/styles/challenges.css" rel="stylesheet">
      Some link
    </link>
  </head>
  <body class="no-js">
    <div class="main-wrapper" role="main">
      Some div
    </div>
    <br />
  </body>
</html>
*/

console.log()
console.log(builder.str())

builder.build(xml => {
  // @ts-ignore
  xml.report(() => {
    // @ts-ignore
    xml.name(() => 'Annual Report')
    times(12, (i: number) => {
      // @ts-ignore
      xml.amounts({ month: i + 1 }, () => {
        // @ts-ignore
        xml.expenses(rand(1000))
        // @ts-ignore
        xml.revenue(rand(1000))
      })
    })
  })
})

console.log()
console.log(builder.str())

/*
<report>
  <name>
    Annual Report
  </name>
  <amounts month="1">
    <expenses>
      594
    </expenses>
    <revenue>
      993
    </revenue>
  </amounts>
  <amounts month="2">
    <expenses>
      299
    </expenses>
    <revenue>
      546
    </revenue>
  </amounts>
  <amounts month="3">
    <expenses>
      356
    </expenses>
    <revenue>
      768
    </revenue>
  </amounts>
  <amounts month="4">
    <expenses>
      323
    </expenses>
    <revenue>
      612
    </revenue>
  </amounts>
  <amounts month="5">
    <expenses>
      319
    </expenses>
    <revenue>
      233
    </revenue>
  </amounts>
  <amounts month="6">
    <expenses>
      939
    </expenses>
    <revenue>
      684
    </revenue>
  </amounts>
  <amounts month="7">
    <expenses>
      83
    </expenses>
    <revenue>
      485
    </revenue>
  </amounts>
  <amounts month="8">
    <expenses>
      33
    </expenses>
    <revenue>
      230
    </revenue>
  </amounts>
  <amounts month="9">
    <expenses>
      591
    </expenses>
    <revenue>
      165
    </revenue>
  </amounts>
  <amounts month="10">
    <expenses>
      9
    </expenses>
    <revenue>
      4
    </revenue>
  </amounts>
  <amounts month="11">
    <expenses>
      155
    </expenses>
    <revenue>
      120
    </revenue>
  </amounts>
  <amounts month="12">
    <expenses>
      866
    </expenses>
    <revenue>
      127
    </revenue>
  </amounts>
</report>
*/
